File: LeetCode/LeetCode Practice Problems/1079- Letter Tile Possibilities.cpp
class Solution {
public:
    int ans = -1;
    void dfs(string s,bool vis[])
    {
        ans+=1;
        for(int i=0;i<s.length();i++)
        {
            if(vis[i])
                continue;
            if(i>0 && s[i] == s[i-1] && !vis[i-1])
                continue;
            
            vis[i] = true;
            dfs(s,vis);
            vis[i] = false;
        }
    }
    int numTilePossibilities(string tiles) {
        if(tiles.length() == 0)
            return 0;
        sort(tiles.begin() , tiles.end());
        bool vis[tiles.length()];
        cout<<tiles;
        dfs(tiles,vis);
        return ans;
    }
};

File: LeetCode/LeetCode Practice Problems/1079-Letter Tile Possibilities.java
class Solution {
    public int numTilePossibilities(String tiles) {
        int cnt[] = new int[26];
        for(char c: tiles.toCharArray()) {
            cnt[c - 'A']++;
        }
        
        return dfs(cnt);
    }
    
    public int dfs(int cnt[]) {
        int ans = 0;
        for(int i=0;i<26;i++) {
            if(cnt[i] == 0) continue;
            cnt[i]--;
            ans++;
            ans += dfs(cnt);
            cnt[i]++;
        }
        return ans;
    }
}
